py-waves: new package

kbrindley opened the pull request:
Adds package WAVES as py-waves: https://github.com/lanl-aea/waves/

> [WAVES](https://lanl-aea.github.io/waves/index.html) (LANL code C23004) is a computational science and engineering workflow tool that integrates parametric studies with traditional software build systems.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

kbrindley committed:
WIP: draft spack recipe

kbrindley committed:
WIP: more complete dependency list. spack reports version conflicts when using the minimum pins of waves pyproject.toml. Start with no minimum pins

kbrindley committed:
WIP: draft test spec

kbrindley committed:
WIP: guessing at the develop branch build requirements

kbrindley committed:
WIP: spack doesn't complain about build minimum pins

kbrindley committed:
FEAT: backward compatibility 'this or newer' dependency specs

kbrindley committed:
WIP: mostly passing tests when requiring runtime dependencies at test time. draft outline for documentation build dependencies

kbrindley committed:
WIP: draft install documentation from an scons build

kbrindley committed:
WIP: spack complains that some docs dependencies require a lower version of sphinx

kbrindley committed:
WIP: sphinx 7.1 required for maximum_signature_line_length support, but I think HTML docs will still build without this feature and without changes to sphinx configuration. Looks like there is no copy button package in spack, so the docs won't build

kbrindley committed:
WIP: more permissible documentation build dependencies. Documentation builds, but may have copybutton bugs and less ideal API formatting. The actual install to package function does not work

kbrindley committed:
WIP: README and pyproject.toml didn't make it into the package, but the 'recipe-matrix' directory no longer shows up in the spack site-packages directory. Progress?

kbrindley committed:
WIP: experimenting with MANIFEST patches to get the pyproject.toml (and therefore the pytest ini options) into the spack package. Turns out this is broken in the upstream, too.

kbrindley committed:
WIP: confirmed that the documentation is included correctly. The 'echo' commands didn't work because spack quotes the shell redirection and bash treats it like a string literal

kbrindley committed:
WIP: add pytest options. Currently testing in source directory, so picking up defaults from pyproject.toml, which is not part of the as-installed package. Figure out how to test the as-installed package

kbrindley committed:
WIP: add parallel test execution

kbrindley committed:
WIP: attempt a version conditional salib spec. Match to upstream spec even though upstream spec is not quite right

kbrindley committed:
DOC: inline dev note for a future man page installation improvement

kbrindley committed:
WIP: use Python functions in copy operations to avoid OS/shell specific build instructions. draft install time creation of share data files which can't be correctly located in a setuptools build/install following the upstream's package_documentation.py logic

kbrindley committed:
WIP: try to find installed package for testing

kbrindley committed:
WIP: trying to use spack provided site directories package instead of searching for it

kbrindley committed:
WIP: for some reason we aren't picking up a new enough version of python for the exists option?

kbrindley committed:
FEAT: installation matching conda package when performed with 'py-waves+docs'

kbrindley committed:
MAINT: replace commented out dependency executable code with todo spec comments

kbrindley committed:
WIP: replace commented out keyword arguemnts with try:except statements. Better to figure out why an older version of Python than expected is under use

kbrindley committed:
DOC: future improvement to use newer python version. remove comment about using spack install_tree, which doesn't seem to have an ignore option

kbrindley committed:
MAINT: spack style black/flake8 updates

kbrindley committed:
FEAT: patch the upstream MANIFEST to include data files. Most important is pyproject.toml for the pytest default options

kbrindley committed:
MAINT: update developer notes

kbrindley committed:
WIP: do some version inspection to decide if source edits are necessary and skip known (but mostly harmless) unit test errors. Flesh out the install override maintainer dev note

kbrindley committed:
WIP: ignore the salib unit tests on WAVES versions known to have a too low salib spec. I seem to have broken the import tests somehow?

kbrindley committed:
WIP: always build the documentation so we can match the conda-build recipe and upstream package_documentation.py script more closely

kbrindley committed:
MAINT: all build requirements are the same now (no special +docs handling) so specify in a single block

kbrindley committed:
WIP: use a more spack package build direct spec inspection solution

kbrindley committed:
MAINT: all build requirements are the same now (no special +docs handling) so specify in a single block

kbrindley committed:
WIP: whitelist the public API for import testing

kbrindley committed:
WIP: patch broken install-time test imports. Unfortunate side effect is that this also patches the functional spack test run py-waves behavior

kbrindley committed:
MAINT: remove unused import

kbrindley commented:
Style check says the following. I used the spack generated template recipe. I checked the header against SCons and numpy recipes and they look about the same. Not sure how to fix.

> ==> Error: 1 improperly licensed files
>   files with wrong SPDX-License-Identifier:   0
>   files without license in first 6 lines:     1
>   files not containing expected license:      0

kbrindley committed:
MAINT: remove future release dependency until WAVES v0.12.9 is officially released and SALib 1.4.6 spack build is available.

kbrindley committed:
MAINT: add docstring to temporary test redifinition to satisfy CI tests

kbrindley commented:
I added the test function docstring, but the test imports function is empty because there's a bug in the default test definition

> PKG-PROPERTIES: 2 issues found
> 1. Package py-waves has test method with empty or missing docstring.
>     Docstrings are used as descriptions in test outputs.
>     Add a concise summary to the following method in 'spack.pkg.builtin.py-waves':
>     test_imports
> 2. Package py-waves has empty or missing test method.
>     Implement or remove the following method from 'spack.pkg.builtin.py-waves': test_imports

tldahlgren commented on var/spack/repos/builtin/packages/py-waves/package.py:
```diff
@@ -0,0 +1,138 @@
+
+    license("BSD-3-Clause", checked_by="kbrindley")
+
+    version("develop", branch="main", get_full_repo=True)
```
Is there a reason the version name doesn't match the branch? We generally don't do that anymore (see https://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison).

kbrindley replied:
WAVES uses a single development branch "main". I was trying to match the idea of a "develop" installation found in other projects. It looks like the preference is to match the "infinity" version to the branch name? I will make that change.

kbrindley replied:
Hm. I was trying to use the ``self.spec.version.isdevelop()`` method to decide when to set the ``SETUPTOOLS_SCM_PRETEND_VERSION`` vs requiring Git and setuptools_scm to decide the version during an "infinity" build.

Is there a version name preference for infinity builds? Should I find another way to check build type or maybe set the infinity build version to ``main``?

I suppose I can also just remove the infinity build option until this PR and my understanding of spack guidance is more mature.

tldahlgren replied:
Packages that have branches with the name `main` use `main` as the version. This will ensure the `main` branch is always newer than any numbered version branch.

kbrindley replied:
I only see two examples of setting ``SETUPTOOLS_SCM_PRETEND_VERSION`` conditionally on the infinity build version. These both use the ``self.spec.version.isdevelop()`` method as copied in this package.

I can't find the docs for this method. Closest I can find is: https://spack.readthedocs.io/en/latest/spack.version.html#spack.version.StandardVersion.isdevelop.

> Triggers on the special case of the @develop-like version.

Is this the best way to check for an infinity version? Should I check as a less than operation instead?

```
if self.spec.version < "stable"
```

I can't pass "main" as a version to ``SETUPTOOLS_SCM_PRETEND_VERSION`` or py-build/setuptools will complain that the version doesn't meet the PEP-440 standard.

-----

```
$ grep -riIE "SETUPTOOLS_SCM_PRETEND_VERSION" var/spack/repos/builtin/packages/py-*
var/spack/repos/builtin/packages/py-geopmdpy/package.py:            env.set("SETUPTOOLS_SCM_PRETEND_VERSION", self.version)
var/spack/repos/builtin/packages/py-geopmpy/package.py:            env.set("SETUPTOOLS_SCM_PRETEND_VERSION", self.version)
```

tldahlgren replied:
Are you saying the current check -- `if not self.spec.version.isdevelop()` -- doesn't work for version `main`?

kbrindley replied:
I wasn't sure if it would work. I also wasn't sure if there was a spack preferred semantic for checking infinity versions. I can only find a few examples of the ``isdevelop`` method in use.

From your python session below it looks like this will work. Thanks!

tldahlgren reviewed (changes requested):
Confirmed the homepage, main branch existence, version sha256, the three requirements and ten dependencies (explicitly listed in `pyproject.toml`).

Is there a reason not to call the version associated with the branch `main`?

kbrindley committed:
MAINT: match infinity version to upstream branch name

tldahlgren commented on var/spack/repos/builtin/packages/py-waves/package.py:
```diff
@@ -0,0 +1,138 @@
+    def setup_build_environment(self, env):
+        env.set("PREFIX", self.prefix)
+        env.set("PKG_NAME", "waves"),
+        if not self.spec.version.isdevelop():
```
This returns `False` if the version is `main`:

```
$ spack-python
>>> import spack.version
>>> spack.version.Version("main").isdevelop()
True
```

kbrindley replied:
Thanks! I hadn't seen the ``spack-python`` command yet. I'll keep the current ``isdevelop()`` check.

kbrindley commented on var/spack/repos/builtin/packages/py-waves/package.py:
```diff
@@ -0,0 +1,138 @@
+            pytest = which("pytest")
+            pytest(*custom_arguments)
+
+    # TODO: Remove when the default PythonPackage test_imports is fixed
+    # https://github.com/spack/spack/issues/48527
+    # https://github.com/spack/spack/issues/45801
+    # https://github.com/spack/spack/pull/48394
+    def test_imports(self):
+        """Skip import tests while waiting for spack PythonPackage bugfix"""
+        pass
```
I think the package audit precheck is complaining about the ``pass`` in this function. I struggled with the install time import tests until I found this issue (https://github.com/spack/spack/issues/45801) and PR (https://github.com/spack/spack/pull/48394).

Should I keep this and find a way to satisfy the package audit precheck? Should I remove it and allow the install tests to be broken until a patch is merged?

edit: the check in question https://github.com/spack/spack/actions/runs/14182716622/job/39732231737?pr=49805

```
1. Package py-waves has empty or missing test method.
    Implement or remove the following method from 'spack.pkg.builtin.py-waves': test_imports
```

kbrindley replied:
I removed. I guess it's ok to use the existing infrastructure and just cite the open issue/PR.

kbrindley commented on var/spack/repos/builtin/packages/py-waves/package.py:
```diff
@@ -0,0 +1,138 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
```
I think the precheck doesn't like this license identifier? This was the spack generated template and the identifier seems to match other packages. I'm not sure how to address the style precheck.

https://github.com/spack/spack/actions/runs/14182716622/job/39732230063?pr=49805

```
/home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-waves/package.py: License not found in first 6 lines
```

kbrindley replied:
updated to match all other python packages. style check passes now.

kbrindley committed:
FEAT: add newly released versions

kbrindley committed:
MAINT: match the header from every other python package recipe

kbrindley committed:
MAINT: remove empty test imports function. Install time import tests will fail until https://github.com/spack/spack/issues/45801 and https://github.com/spack/spack/pull/48394 are addressed

kbrindley commented:
> Is there a reason not to call the version associated with the branch `main`?

@tldahlgren I fixed the ``main`` branch name inconsistency and addressed the style and audit prechecks.
